### PR TITLE
Adding in filtering for dashboard table

### PR DESCRIPTION
**Pull request by jamiepollock**
This PR should cover basic filtering of data outlined in #6 .

A user can search invariant exact for a name, email user type and also whether a user is active or inactive.

The case doesn't matter but it must be the full value.

I've tested this against Umbraco v7.4.3 with a few 1000 users. To create so many users I setup an UmbracoApiController like so to quickly create multiple users of a specific type. I was initially a bit concerned with how it would work with large numbers of users to bring back and filter by Linq but it's not so bad actually. There is a short delay but nothing world-breaking.

_Note:_ This doesn't currently cover power words which allow users to do stuff like `is:admin`. I'd consider this a stage 2 unless otherwise agreed.

``` c#
using System;
using System.Net;
using System.Net.Http;
using System.Web.Http;
using Umbraco.Web.WebApi;

namespace DemoApp.Controllers
{
    public class UserTestApiController : UmbracoApiController
    {
        [HttpGet]
        public HttpResponseMessage Create(int number = 50, string userTypeAlias = "admin")
        {
            if (number < 1)
            {
                return Request.CreateValidationErrorResponse("number must be higher than 0");
            }


            var userType = Services.UserService.GetUserTypeByAlias(userTypeAlias);

            if (userType == null)
            {
                return
                    Request.CreateValidationErrorResponse("Unable to find a user type with the alias, " + userTypeAlias);
            }

            var count = 0;

            for (var index = 0; index <= number; index++)
            {

                try
                {
                    var userName = string.Format("User{0}-{1:yy-mm-dd}", index, DateTime.Now);
                    var email = string.Format("{0}@foo.com", userName.Replace("-", string.Empty));

                    var exists = Services.UserService.Exists(userName);

                    if (exists == false)
                    {
                        var user = Services.UserService.CreateUserWithIdentity(userName, email, userType);

                        Services.UserService.Save(user);
                    }
                }
                catch (Exception ex)
                {
                    return Request.CreateErrorResponse(HttpStatusCode.InternalServerError, ex);
                }
            }


            return Request.CreateResponse(HttpStatusCode.OK,
                string.Format("{0} successfully {1} create", number, userTypeAlias));
        }

    }
}
```

### CODE REVIEW

**Comment by mattbrailsford** [comment]
Wonder if an experiment might be worthwhile to test setting up another lucene index to search against?

**Comment by jamiepollock** [comment]
@mattbrailsford sounds like an interesting idea. I've not created a custom Lucene index before might not be something I can get done tomorrow but could certainly look into it.

Obviously we'd be sorting and searching against just the data required for this list. Name, Email, User Type & Active. Aside from email everything is pretty standard any not too out of the ordinary of what you'd find in typical a Lucene index.

What do you think @leekelleher?

**Comment by leekelleher** [comment]
To be honest, I think having an Examine index purely for the filtering
seems overkill.  Unless we were thinking of other cases for it? Even then,
is that outside the scope of this bulk admin tool?

@jamiepollock Out of curiosity, what's the most users you've seen in the
back-office?

On Carlsberg, we've had around 150 users, which doesn't seem to crazy for a
Linq query.

**Comment by jamiepollock** [comment]
I can take a headcount tomorrow from our site we do for Tarmac but yea 150-200 tops.

**Comment by jamiepollock** [comment]
@leekelleher I'm wondering about partial matches on certain fields. This kinda thing might make sense for Name & Email but not so much for User Type (maybe partial is okay?) & Active. What do you think?

**Comment by leekelleher** [comment]
Partial matches sounds good. We had a case where we needed to bulk edit all the users from an external agency, where they all had the same domain in their email address.

**Comment by jamiepollock** [comment]
yea, I'm thinking that's the best shout for obvious reason Active/Inactive shouldn't be in there.

I'm hesitant to going much further with this because searching logic of stuff like "admin inactive" might get messy.

Users could search admin then sort by Active to find the users who are inactive maybe?

To confirm Name, Email & User Type are going to be partial search :)

**Comment by jamiepollock** [comment]
@leekelleher Okay, looks like that's ticking along nicely. :)

**Comment by jamiepollock** [comment]
I can't help but feel that filtering function is really messy and barely maintainable. If you have any suggestions I'm all ears.

**Comment by leekelleher** [comment]
@jamiepollock I'll put an inline comment on the code.

**Comment by jamiepollock** [comment]
@leekelleher made a bit of a change in this code. This is a bit of a change to how things work, so review with caution.

I personally find this a bit more readable but you may disagree, which is totally fine! _We can always revert™._

File: src/Our.Umbraco.BulkUserAdmin/Web/Controllers/BulkUserAdminApiController.cs [hunk 1]
```diff
@@ -14,35 +15,62 @@ public class BulkUserAdminApiController : UmbracoAuthorizedApiController
...
+            {
+                Func<BulkUserListItemModel, bool> partialMatchOnFields = item => new[] {
+                    item.Name,
+                    item.Email,
+                    item.UserType
+                }.Any(x => x.IndexOf(f, StringComparison.InvariantCultureIgnoreCase) > -1);
+
+                Func<BulkUserListItemModel, bool> exactMatchOnActive = item => string.Equals(f, item.Active ? FilterTermActive : FilterTermInactive, StringComparison.OrdinalIgnoreCase);
+
+                var filteredItems = items.Where(item => partialMatchOnFields(item) || exactMatchOnActive(item));
```
Comment (leekelleher) [inline src/Our.Umbraco.BulkUserAdmin/Web/Controllers/BulkUserAdminApiController.cs:57]: @jamiepollock Do you mean that this code felt unmaintainable?

I think this is fine, (at least better than the previous commit).

**Comment by leekelleher** [comment]
@jamiepollock Looks good to me.  I'm happy to merge in whenever you're happy.

File: src/Our.Umbraco.BulkUserAdmin/Web/Controllers/BulkUserAdminApiController.cs (hunk 1 above)
Comment (jamiepollock) [inline src/Our.Umbraco.BulkUserAdmin/Web/Controllers/BulkUserAdminApiController.cs:57]: yea, previously with the MAHUSSIF Where clause. Needed to add a class in so I could do a Func<BulkUserListItemModel, bool> against it.

**Comment by jamiepollock** [comment]
@leekelleher yea, I'm happy with it if you're happy with the code changes I've made.

I can't think of too much else to add to this package now? Surely we should be close to getting this released ;) no pressure! :D

**Comment by leekelleher** [comment]
Merged in! Thanks @jamiepollock!

I'll add you as a contributor on the repo, gives you write access. If any new features do come up, we'll do as we do on Ditto, branch it off and do an internal PR for discussions, etc.

**Comment by jamiepollock** [comment]
@leekelleher well this is all very exciting! :)

Thanks Lee

**Pull request merged**